File: JAVA/java/src/CreateBinaryTreeFromDescriptions.java
import java.util.*;

public class CreateBinaryTreeFromDescriptions {
    public TreeNode createBinaryTree(int[][] descriptions) {
        // Step 1: Organize data
        Map<Integer, List<int[]>> parentToChildren = new HashMap<>();
        Set<Integer> allNodes = new HashSet<>();
        Set<Integer> children = new HashSet<>();

        for (int[] desc : descriptions) {
            int parent = desc[0];
            int child = desc[1];
            int isLeft = desc[2];

            // Store child information under parent node
            if (!parentToChildren.containsKey(parent)) {
                parentToChildren.put(parent, new ArrayList<>());
            }
            parentToChildren.get(parent).add(new int[] { child, isLeft });
            allNodes.add(parent);
            allNodes.add(child);
            children.add(child);
        }

        // Step 2: Find the root
        int rootVal = 0;
        for (int node : allNodes) {
            if (!children.contains(node)) {
                rootVal = node;
                break;
            }
        }

        // Step 3 & 4: Build the tree using DFS
        return dfs(parentToChildren, rootVal);
    }

    // DFS function to recursively build binary tree
    private TreeNode dfs(Map<Integer, List<int[]>> parentToChildren, int val) {
        // Create new TreeNode for current value
        TreeNode node = new TreeNode(val);

        // If current node has children, recursively build them
        if (parentToChildren.containsKey(val)) {
            for (int[] childInfo : parentToChildren.get(val)) {
                int child = childInfo[0];
                int isLeft = childInfo[1];

                // Attach child node based on isLeft flag
                if (isLeft == 1) {
                    node.left = dfs(parentToChildren, child);
                } else {
                    node.right = dfs(parentToChildren, child);
                }
            }
        }

        return node;
    }

    // Helper function to print the tree (in-order traversal)
    private void printTree(TreeNode root) {
        if (root == null) {
            return;
        }
        printTree(root.left);
        System.out.print(root.val + " ");
        printTree(root.right);
    }

    // Helper function to print the tree (level-order traversal)
    private void printTreeLevelOrder(TreeNode root) {
        if (root == null) {
            return;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            System.out.print(node.val + " ");
            if (node.left != null) {
                queue.add(node.left);
            }
            if (node.right != null) {
                queue.add(node.right);
            }
        }
    }

    public static void main(String[] args) {
        CreateBinaryTreeFromDescriptions treeBuilder = new CreateBinaryTreeFromDescriptions();

        int[][] descriptions = {
                {20, 15, 1},
                {20, 17, 0},
                {50, 20, 1},
                {50, 80, 0},
                {80, 19, 1}
        };

        TreeNode root = treeBuilder.createBinaryTree(descriptions);

        // Print the tree (in-order traversal)
        System.out.print("In-order traversal: ");
        treeBuilder.printTree(root);
        System.out.println();

        // Print the tree (level-order traversal)
        System.out.print("Level-order traversal: ");
        treeBuilder.printTreeLevelOrder(root);  // Expected output: 50 20 80 15 17 19
        System.out.println();
    }
}

File: python/Create Binary Tree From Descriptions.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class CreateBinaryTreeFromDescriptions:
    def create_binary_tree(self, descriptions):
        from collections import defaultdict

        # Step 1: Organize data
        parent_to_children = defaultdict(list)
        all_nodes = set()
        children = set()

        for parent, child, is_left in descriptions:
            parent_to_children[parent].append((child, is_left))
            all_nodes.add(parent)
            all_nodes.add(child)
            children.add(child)

        # Step 2: Find the root
        root_val = next(node for node in all_nodes if node not in children)

        # Step 3 & 4: Build the tree using DFS
        return self.dfs(parent_to_children, root_val)

    def dfs(self, parent_to_children, val):
        # Create new TreeNode for current value
        node = TreeNode(val)

        # If current node has children, recursively build them
        for child, is_left in parent_to_children.get(val, []):
            if is_left == 1:
                node.left = self.dfs(parent_to_children, child)
            else:
                node.right = self.dfs(parent_to_children, child)

        return node

    # Helper function to print the tree (in-order traversal)
    def print_tree_in_order(self, root):
        if not root:
            return
        self.print_tree_in_order(root.left)
        print(root.val, end=' ')
        self.print_tree_in_order(root.right)

    # Helper function to print the tree (level-order traversal)
    def print_tree_level_order(self, root):
        if not root:
            return
        from collections import deque
        queue = deque([root])
        while queue:
            node = queue.popleft()
            print(node.val, end=' ')
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

if __name__ == "__main__":
    tree_builder = CreateBinaryTreeFromDescriptions()

    descriptions = [
        [20, 15, 1],
        [20, 17, 0],
        [50, 20, 1],
        [50, 80, 0],
        [80, 19, 1]
    ]

    root = tree_builder.create_binary_tree(descriptions)

    # Print the tree (in-order traversal)
    print("In-order traversal: ", end='')
    tree_builder.print_tree_in_order(root)
    print()

    # Print the tree (level-order traversal)
    print("Level-order traversal: ", end='')
    tree_builder.print_tree_level_order(root)
    print()
